### PR TITLE
SOLR-12398: Add Heatmap option to JSON Facet API.

### PR DESCRIPTION
**Pull request by dsmiley**
* moved the preponderance of the implementation from SpatialHeatmapFacets (used by SimpleFacets) into the new API.
* no Solr ref guide changes

### CODE REVIEW

File: solr/test-framework/src/java/org/apache/solr/BaseDistributedSearchTestCase.java [hunk 1]
```diff
@@ -606,6 +608,10 @@ protected QueryResponse query(boolean setDistribParams, SolrParams p) throws Exc
     final QueryResponse controlRsp = controlClient.query(params);
     validateControlData(controlRsp);
 
+    if (shardCount == 0) {//mostly for temp debugging
+      return controlRsp;
+    }
+
```
Comment (dragonsinth) [inline solr/test-framework/src/java/org/apache/solr/BaseDistributedSearchTestCase.java]: this supposed to land?

File: solr/test-framework/src/java/org/apache/solr/BaseDistributedSearchTestCase.java [hunk 2]
```diff
@@ -488,7 +488,9 @@ protected void index(Object... fields) throws Exception {
    */
   protected void indexDoc(SolrInputDocument doc) throws IOException, SolrServerException {
     controlClient.add(doc);
-
+    if (shardCount == 0) {//mostly for temp debugging
+      return;
+    }
```
Comment (dragonsinth) [inline solr/test-framework/src/java/org/apache/solr/BaseDistributedSearchTestCase.java]: this supposed to land?

**Review by dragonsinth — COMMENTED** [review]
Sweet! Let me run through some tests tomorrow.

File: solr/test-framework/src/java/org/apache/solr/BaseDistributedSearchTestCase.java (hunk 1 above)
Comment (dsmiley) [reply solr/test-framework/src/java/org/apache/solr/BaseDistributedSearchTestCase.java]: yes and the next diff too.  This allows you to set numShards=0 temporarily for debugging and and thus only have the "control client".  Granted it defeats the purpose of using BaseDistributedSearchTestCase but it's only meant to temporarily remove distracting aspects while some bug is being worked out that is not related to distributed search.

**Review by dragonsinth — APPROVED** [review]

**Comment by dragonsinth** [comment]
LGTM passes our internal tests as well

**Comment by dragonsinth** [comment]
Thank you David!

**Pull request merged**